src/object-to-ast.ts

```diff
@@ -0,0 +1,48 @@
+import type babelCore from '@babel/core'
+
+module.exports = objectToAST
+
+type ObjectToASTOptions = {
+  babel: typeof babelCore
+  fileOptions: babelCore.TransformOptions
+}
+
+export default function objectToAST(
+  object: unknown,
+  {babel, fileOptions}: ObjectToASTOptions,
+) {
+  const stringified = stringify(object)
+  const variableDeclarationNode = babel.template(`var x = ${stringified}`, {
+    preserveComments: true,
+    placeholderPattern: false,
+    ...fileOptions.parserOpts,
+    sourceType: 'module',
+  })()
+
+  // @ts-expect-error "field is missing in declaration, but available at runtime?"
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+  return variableDeclarationNode.declarations[0].init as babelCore.Node
+}
+
+function stringify(object: unknown) {
+  let str = JSON.stringify(object, stringifyReplacer)
+
+  // TODO: TS says this is unnecessary... remove it?
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (str === undefined) {
+    str = 'undefined'
+  }
```

Comment:
According to TS `JSON.stringify()` won't return `undefined` so this is no necessary.  I was going to run some quick tests to see if that's true, but if somebody knows one way or the other, happy to hear it.

Reply:
Seems like `stringify` always takes an object, so I think we can remove the check if `str` is undefined, unless the `object` itself can be undefined?

Reply:
My assumption is the intent here was to protect against `stringify(undefined)` returning undefined, but TS isn't flagging that as an option.  I want to just write a quite test to check what it would do in that case.

Reply:
Should we then not check if `object` is undefined?

Reply:
I think the implication is:

```ts
stringify(undefined) === ''
stringify(null) === 'null'
stringify('value') === '"value"'
stringify({ key: 'value' }) === '{ "key": "value" }'
```
i.e. you always get a string, regardless of the input. I just don't know if this is true and want to check before ripping the safety check out.

Reply:
I see! Thanks for clarifying its purpose :blue_heart: :pray: :smiling_face_with_three_hearts:

Reply:
The check is needed, I don't know why your setup reports it as being false, the [online repl][1] doesn't.

[MDN also mentiones][2] that `JSON.stringify` can return `undefined`.

> ... JSON.stringify() can return undefined when passing in "pure" values like JSON.stringify(function(){}) or JSON.stringify(undefined).

And the most basic check that you can do to confirm this, is running this in either your browser or node:

```
JSON.stringify(undefined) === undefined
```

**TLDR:** I wouldn't remove that check. Your setup seems to be messing up.

[1]: https://www.typescriptlang.org/play?ts=4.1.3#code/GYVwdgxgLglg9mABAZygJxmA5jYBPACjgCMArAU2gC5FwBrMOAdzAEpEBvAKEUQBtyUFOkQBeRACkAygHkAcgDpUGbLkIkK0ADTCVOfACVyABz4BDCOTSsuPRLkQFlY0ePAATcsEzl37bry8zuIA5B5ePu4hdgC+dnZogiBoSMoKiaYW5AR2vAD0AEQA+kUAYgCqcgDCACoAkvJFUjUAggY1JQQKAFQA-KwlFdX1jQCicgAiJQV5WFq5iKCQsAhGmZZo87w2dkvQ8KnomPp4a+YbBEV05Hg0ysc6AG5mfCDkNPSMLP4LDgRQeGM5DgwEQz1e5BcoT2KzAIR+gUCiSgySQAANBpVag05E1Wu0SgASDjgt4KKBwKRHbAEVgxTHDHFFcZTIpohZxRELZGosEvN6xWy8GEHRbgfarEznKyXAC2ZigEAAFndqXNEMYAIyqvQIpFJFIazXpKVZAh5AA6Fpm6pCBXhJvW2UtFrAsx0IVd8MFcS4EAQqF0x1wMF8YiDqnwBHC3jAvlYAG4uEA

[2]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description

Reply:
the setup is `kcd-scripts`, so not sure where it's getting it's information from

Reply:
so TS itself isn't reporting the issue, eslint is.  If you hover over `str` in the online repl, it shows the type as just `string` instead of `string | undefined`:

![image](https://user-images.githubusercontent.com/23029903/106738564-2ea50580-666c-11eb-915e-20f23d8b6eb1.png)

There is an [open TS issues about this](https://github.com/microsoft/TypeScript/issues/18879) and [an unmerged fix for it](https://github.com/microsoft/TypeScript/pull/38574).

I'm going to update the TODO comment to reference that issue and keep the `eslint-disable-next-line` as it's definitely not something we want to remove.